### PR TITLE
CLOUDSTACK-10436:remind users to use correct permission for tmp dir and fixed an NPE

### PR DESCRIPTION
**Pull request by lujiefsi**
### Description

I think there is a potential securiry issuse in createUniqDir#JavaStorageLayer

 

    public File createUniqDir() {
        String dirName = System.getProperty("java.io.tmpdir");
        if (dirName != null) {
            File dir = new File(dirName);
            if (dir.exists()) {
                String uniqDirName = dir.getAbsolutePath() + File.separator + UUID.randomUUID().toString();
                if (mkdir(uniqDirName)) {
                    return new File(uniqDirName);
                }
            }
        }
        return null;
    }
So if a user do not specify the "java.io.tmpdir",  we will use the default tmp dir "/tmp" whose mode is 777. Even users specify the "java.io.tmpdir", they may also forget to set the file mode as 700, hence the tmp dir is still worldreadable.  Our code then create create UniqDir  in tmp dir, and it is also worldreadable.

 

createUniqDir will be called by swiftUploadMetadataFile and registerTemplateOnSwift in NfsSecondaryStorageResource. Hence the MetadataFile will be written in tmp dir and is also worldreadable. 

 

Hum, assums that cloudstack runs on  a server as root, and there are also some regualr users on this  server. These regualr users (they even do not belong to cloudstack group) can read these medadatafile, which is undesirable.

 

I think this is similar to https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-8908 . 

This PR will remind users to use correct permission for tmp dir.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by rhtyd** [comment]
Hi @lujiefsi have you reproduced any issue on management server? I maybe wrong but many of the template related registration and operations actually run in ssvm which is not generally accessible.

**Comment by lujiefsi** [comment]
Hi @rhtyd  Yes, most  the template related registration and operations actually run in ssvm which is not generally accessible.  what I concern is below code:

 `   protected boolean swiftUploadMetadataFile(SwiftTO swift, File srcFile, String containerName, String uniqueName) throws IOException {

        File uniqDir = _storage.createUniqDir();
        String metaFileName = uniqDir.getAbsolutePath() + File.separator + _tmpltpp;
        _storage.create(uniqDir.getAbsolutePath(), _tmpltpp);// metafile is created

        long virtualSize = getVirtualSize(srcFile, getTemplateFormat(srcFile.getName()));

        File metaFile = swiftWriteMetadataFile(metaFileName, uniqueName, srcFile.getName(), srcFile.length(), virtualSize);///metafile is written with uniquename,filename,size,virtualsize

        SwiftUtil.putObject(swift, metaFile, containerName, _tmpltpp);
        metaFile.delete();
        uniqDir.delete();

        return true;
    }`

are the uniquename,filename,size and virtualsize  sensitive? If not, that's ok. But i still think we need follow The Principle of Least Privilege (PoLP), see https://www.cyberark.com/what-is/least-privilege/

I also still have another concern, that is if others take JavaStorageLayer as library for future, thay maybe not aware that the tmp dir created by createUniqDir is worldreadable.


Meanwhile, i give a new commit to fix the potential NPE when the tmp dir does not exist.

**Comment by DaanHoogland** [comment]
@lujiefsi we should not encourage people to make '/tmp/' have mode 700. That is not a valid mode for /tmp on a unix system. it must have 1777 as mode. You could make sure "java.io.tmpdir" is a subdir of /tmp and it has good rights, but I'm -1 as is.

**Comment by lujiefsi** [comment]
@DaanHoogland Yes, we should not encourage people to make '/tmp/' have mode 700. But we still need to encourage users to override "/tmp" with "java.io.tmpdir". repush the new commit,

**Comment by DaanHoogland** [comment]
> @DaanHoogland Yes, we should not encourage people to make '/tmp/' have mode 700. But we still need to encourage users to override "/tmp" with "java.io.tmpdir". repush the new commit,

ok, I can live with it this way ;) (:+1: )

**Review by DaanHoogland — APPROVED** [review]
code lgtm

**Comment by DaanHoogland** [comment]
@rhtyd do you approve?
@lujiefsi have you tested? do we need more testing?

**Comment by lujiefsi** [comment]
> @rhtyd do you approve?
> @lujiefsi have you tested? do we need more testing?

YES, I have tested this patch!

**Comment by DaanHoogland** [comment]
can @lujiefsi get a seconf LGTM (@rhtyd @andrijapanicsb @GabrielBrascher ... and others)?

File: core/src/main/java/com/cloud/storage/JavaStorageLayer.java [hunk 1]
```diff
@@ -217,6 +229,19 @@ public boolean mkdirs(String path) {
...
 
+    public boolean isWorldReadable(File file) throws IOException {
+        Set<PosixFilePermission> permissions;
+        permissions = Files.getPosixFilePermissions(
+            Paths.get(file.getAbsolutePath()));
+
+        for (PosixFilePermission permission:permissions) {
+            if (permission.equals(PosixFilePermission.OTHERS_READ)) {
+                return true;
+            }
```
Comment (sureshanaparti) [inline core/src/main/java/com/cloud/storage/JavaStorageLayer.java:240]: no need of looping through all permissions, can simply check using permissions set contains `PosixFilePermission.OTHERS_READ` or not

```
if (permissions.contains(PosixFilePermission.OTHERS_READ)) {
                 return true;
}
```

Comment (lujiefsi) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java:240]: have simplified it.

**Comment by rhtyd** [comment]
@blueorangutan package

File: core/src/main/java/com/cloud/storage/JavaStorageLayer.java [hunk 2]
```diff
@@ -178,18 +183,25 @@ public long getSize(String path) {
...
 
     @Override
-    public File createUniqDir() {
+    public File createUniqDir() throws IOException {
         String dirName = System.getProperty("java.io.tmpdir");
         if (dirName != null) {
             File dir = new File(dirName);
             if (dir.exists()) {
+                if (isWorldReadable(dir)) {
+                    if (dir.getAbsolutePath().equals("/tmp")) {
```
Comment (GabrielBrascher) [inline core/src/main/java/com/cloud/storage/JavaStorageLayer.java:192]: What do you think of extracting `"/tmp"` into a constant in this class?

Comment (lujiefsi) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java:192]: /tmp is the default tmp dir for linux. if we directly tell the user "temp dir is is World Readable", we may mislead user to change the permission of "/tmp", which is not desirable. So we just tell user "the temp dir is /tmp" and hope they change the default temp dir.

Comment (DaanHoogland) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java:192]: @lujiefsi I think @GabrielBrascher means that we should add a 
```
    private static final String STDTMP = "/tmp";
```
and than 
```suggestion
                    if (STDTMP.equals(dir.getAbsolutePath()) {
```

File: core/src/main/java/com/cloud/storage/JavaStorageLayer.java [hunk 3]
```diff
@@ -178,18 +183,25 @@ public long getSize(String path) {
...
+                        s_logger.warn("The temp dir is /tmp");
```
Comment (DaanHoogland) [inline core/src/main/java/com/cloud/storage/JavaStorageLayer.java:193]: and than here of course
```suggestion
                        s_logger.warn(String.format("The temp dir is %s", STDTMP);
```

File: core/src/main/java/com/cloud/storage/JavaStorageLayer.java (hunk 2 above)
Comment (GabrielBrascher) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java:192]: Sorry for not being clear about the idea, @lujiefsi.
Good that @DaanHoogland explained it better than me :slightly_smiling_face:

Comment (lujiefsi) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java:192]: sorry for my mis-undstand

**Comment by nvazquez** [comment]
Hi @lujiefsi looks like all the reviews have been addressed, can you please mark them as resolved?
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]
LGTM

**Comment by nvazquez** [comment]
@blueorangutan test keepEnv

File: core/src/main/java/com/cloud/storage/JavaStorageLayer.java [hunk 4]
```diff
@@ -21,15 +21,21 @@
...
+import java.util.Set;
 import java.util.UUID;
 
 import javax.naming.ConfigurationException;
+import org.apache.log4j.Logger;
 
 public class JavaStorageLayer implements StorageLayer {
-
+    private static final Logger s_logger = Logger.getLogger(JavaStorageLayer.class);
+    private static final String STDTMP = "/tmp";
```
Comment (sureshanaparti) [inline core/src/main/java/com/cloud/storage/JavaStorageLayer.java:38]:
```suggestion
    private static final String STD_TMP_DIR_PATH = "/tmp";
```

Comment (lujiefsi) [reply core/src/main/java/com/cloud/storage/JavaStorageLayer.java]: good point

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Pull request merged**